presto-main/src/main/java/com/facebook/presto/operator/ExplainAnalyzeOperator.java

```diff
@@ -0,0 +1,145 @@
+            this.metadata = requireNonNull(metadata, "metadata is null");
+        }
+
+        @Override
+        public List<Type> getTypes()
+        {
+            return ImmutableList.of(VARCHAR);
+        }
+
+        @Override
+        public Operator createOperator(DriverContext driverContext)
+        {
+            checkState(!closed, "Factory is already closed");
+            OperatorContext operatorContext = driverContext.addOperatorContext(operatorId, planNodeId, ExplainAnalyzeOperator.class.getSimpleName());
+            return new ExplainAnalyzeOperator(operatorContext, queryPerformanceFetcher, metadata);
+        }
+
+        @Override
+        public void close()
+        {
+            closed = true;
+        }
+
+        @Override
+        public OperatorFactory duplicate()
+        {
+            return new ExplainAnalyzeOperatorFactory(operatorId, planNodeId, queryPerformanceFetcher, metadata);
+        }
+    }
+
+    private final OperatorContext operatorContext;
+    private final QueryPerformanceFetcher queryPerformanceFetcher;
+    private final Metadata metadata;
+    private boolean finishing;
+    private boolean outputConsumed;
+
+    public ExplainAnalyzeOperator(OperatorContext operatorContext, QueryPerformanceFetcher queryPerformanceFetcher, Metadata metadata)
+    {
+        this.operatorContext = requireNonNull(operatorContext, "operatorContext is null");
+        this.queryPerformanceFetcher = requireNonNull(queryPerformanceFetcher, "queryPerformanceFetcher is null");
+        this.metadata = requireNonNull(metadata, "metadata is null");
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public List<Type> getTypes()
+    {
+        return ImmutableList.of(VARCHAR);
+    }
+
+    @Override
+    public void finish()
+    {
+        finishing = true;
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return finishing && outputConsumed;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return !finishing;
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        checkState(needsInput());
+
+        // Ignore the input
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        if (!finishing) {
+            return null;
+        }
+        outputConsumed = true;
+        QueryInfo queryInfo = queryPerformanceFetcher.getQueryInfo(operatorContext.getDriverContext().getTaskId().getQueryId());
+        // Skip the output stage, since that's the one that has the ExplainAnalyzeOperator itself
+        String plan = textDistributedPlan(queryInfo.getOutputStage().getSubStages(), metadata, operatorContext.getSession());
+        BlockBuilder builder = VARCHAR.createBlockBuilder(new BlockBuilderStatus(), 1);
+        VARCHAR.writeString(builder, plan);
+        return new Page(builder.build());
```

Comment:
It's not that simple. If I change this to return `null` until the stats are collected, the driver thread will spin unless it also transitions to the blocked state, and there's no way for me to know when to transition out of the blocked state